File: a2ui/validate.go
package a2ui

import "fmt"

// ValidationError describes a single structural problem with an A2UI message or
// surface tree. Its JSON form matches the protocol's standard validation error
// payload, allowing it to be returned to an LLM for self-correction.
type ValidationError struct {
	Code      string `json:"code"`
	SurfaceID string `json:"surfaceId,omitempty"`
	Path      string `json:"path,omitempty"`
	Message   string `json:"message"`
}

// Error implements the error interface.
func (e ValidationError) Error() string {
	if e.Path != "" {
		return fmt.Sprintf("%s at %s (surface %q): %s", e.Code, e.Path, e.SurfaceID, e.Message)
	}
	return fmt.Sprintf("%s (surface %q): %s", e.Code, e.SurfaceID, e.Message)
}

// ToClientError converts the validation error into the ClientError wire form
// that a client sends back to the server.
func (e ValidationError) ToClientError() ClientError {
	return ClientError{Code: e.Code, SurfaceID: e.SurfaceID, Path: e.Path, Message: e.Message}
}

func validationErr(surfaceID, path, format string, args ...any) ValidationError {
	return ValidationError{
		Code:      CodeValidationFailed,
		SurfaceID: surfaceID,
		Path:      path,
		Message:   fmt.Sprintf(format, args...),
	}
}

// ValidateEnvelope checks a single server-to-client envelope for structural
// validity against the supplied catalog. The catalog may be nil to skip
// component/function type checks. It returns all problems found.
func ValidateEnvelope(env Envelope, cat *Catalog) []ValidationError {
	var errs []ValidationError

	if env.Kind() == KindUnknown {
		return []ValidationError{validationErr("", "", "envelope contains no recognized message body")}
	}

	switch env.Kind() {
	case KindCreateSurface:
		m := env.CreateSurface
		if m.SurfaceID == "" {
			errs = append(errs, validationErr("", "/createSurface/surfaceId", "surfaceId is required"))
		}
		if m.CatalogID == "" {
			errs = append(errs, validationErr(m.SurfaceID, "/createSurface/catalogId", "catalogId is required"))
		}
	case KindUpdateComponents:
		m := env.UpdateComponents
		if m.SurfaceID == "" {
			errs = append(errs, validationErr("", "/updateComponents/surfaceId", "surfaceId is required"))
		}
		for i, c := range m.Components {
			base := fmt.Sprintf("/updateComponents/components/%d", i)
			if c.ID == "" {
				errs = append(errs, validationErr(m.SurfaceID, base+"/id", "component id is required"))
			}
			if c.Type == "" {
				errs = append(errs, validationErr(m.SurfaceID, base+"/component", "component type is required"))
				continue
			}
			if cat != nil && !cat.HasComponent(c.Type) {
				errs = append(errs, validationErr(m.SurfaceID, base+"/component", "unknown component type %q", c.Type))
			}
		}
	case KindUpdateDataModel:
		m := env.UpdateDataModel
		if m.SurfaceID == "" {
			errs = append(errs, validationErr("", "/updateDataModel/surfaceId", "surfaceId is required"))
		}
	case KindDeleteSurface:
		m := env.DeleteSurface
		if m.SurfaceID == "" {
			errs = append(errs, validationErr("", "/deleteSurface/surfaceId", "surfaceId is required"))
		}
	}
	return errs
}

// ValidateSurfaceTree checks that a fully assembled surface forms a valid
// component tree: a root must exist, every referenced child must be defined,
// component types must exist in the catalog, and there must be no cycles.
func ValidateSurfaceTree(s *Surface, cat *Catalog) []ValidationError {
	var errs []ValidationError

	if _, ok := s.Root(); !ok {
		errs = append(errs, validationErr(s.ID, "", "surface has no %q component", RootComponentID))
	}

	for id, c := range s.Components {
		if cat != nil && c.Type != "" && !cat.HasComponent(c.Type) {
			errs = append(errs, validationErr(s.ID, "/"+id, "unknown component type %q", c.Type))
		}
		for _, ref := range childRefs(c, cat) {
			if _, ok := s.Components[ref]; !ok {
				errs = append(errs, validationErr(s.ID, "/"+id, "references undefined component %q", ref))
			}
		}
	}

	errs = append(errs, detectCycles(s, cat)...)
	return errs
}

// detectCycles reports components that participate in a child-reference cycle,
// reachable from root.
func detectCycles(s *Surface, cat *Catalog) []ValidationError {
	const (
		white = 0 // unvisited
		gray  = 1 // on the current DFS stack
		black = 2 // fully explored
	)
	color := make(map[string]int, len(s.Components))
	var errs []ValidationError

	var visit func(id string)
	visit = func(id string) {
		c, ok := s.Components[id]
		if !ok {
			return
		}
		color[id] = gray
		for _, ref := range childRefs(c, cat) {
			switch color[ref] {
			case gray:
				errs = append(errs, validationErr(s.ID, "/"+id, "circular reference involving %q", ref))
			case white:
				visit(ref)
			}
		}
		color[id] = black
	}

	if _, ok := s.Root(); ok {
		visit(RootComponentID)
	}
	// Also cover components not reachable from root so cycles elsewhere surface.
	for id := range s.Components {
		if color[id] == white {
			visit(id)
		}
	}
	return errs
}

File: desktop/frontend/src/a2ui-renderer/validate.ts
/**
 * A2UI 开发期结构校验。
 *
 * 对齐 Go 端 a2ui/validate.go：
 *   - ValidateEnvelope — 检查信封消息体结构
 *   - ValidateSurfaceTree — 检查组件树完整性（root 存在、无 dangling ref、无 cycle）
 *
 * 校验失败时 console.error 但不阻断渲染。
 */

import type { Envelope, SurfaceState } from './store'
import { envelopeKind, RootComponentID } from './store'
import type { Catalog } from './catalog'
import { childRefsOf } from './child-refs'

// ── 校验错误 ──────────────────────────────────────

export interface ValidationError {
  code: string
  surfaceId?: string
  path?: string
  message: string
}

function _err(surfaceId: string, path: string, message: string): ValidationError {
  return { code: 'VALIDATION_FAILED', surfaceId, path, message }
}

// ── Envelope 校验 ─────────────────────────────────

/**
 * 校验单个服务器信封的结构合法性。
 */
export function validateEnvelope(env: Envelope, cat?: Catalog): ValidationError[] {
  const errs: ValidationError[] = []

  const kind = envelopeKind(env)
  if (kind === '') {
    return [_err('', '', 'envelope contains no recognized message body')]
  }

  switch (kind) {
    case 'createSurface': {
      const m = env.createSurface!
      if (!m.surfaceId) {
        errs.push(_err('', '/createSurface/surfaceId', 'surfaceId is required'))
      }
      if (!m.catalogId) {
        errs.push(_err(m.surfaceId ?? '', '/createSurface/catalogId', 'catalogId is required'))
      }
      break
    }
    case 'updateComponents': {
      const m = env.updateComponents!
      if (!m.surfaceId) {
        errs.push(_err('', '/updateComponents/surfaceId', 'surfaceId is required'))
      }
      for (let i = 0; i < (m.components?.length ?? 0); i++) {
        const c = m.components[i]
        const base = `/updateComponents/components/${i}`
        if (!c.id) {
          errs.push(_err(m.surfaceId ?? '', base + '/id', 'component id is required'))
        }
        if (!c.type) {
          errs.push(_err(m.surfaceId ?? '', base + '/component', 'component type is required'))
        } else if (cat && !cat.components[c.type]) {
          errs.push(_err(m.surfaceId ?? '', base + '/component', `unknown component type "${c.type}"`))
        }
      }
      break
    }
    case 'updateDataModel': {
      const m = env.updateDataModel!
      if (!m.surfaceId) {
        errs.push(_err('', '/updateDataModel/surfaceId', 'surfaceId is required'))
      }
      break
    }
    case 'deleteSurface': {
      const m = env.deleteSurface!
      if (!m.surfaceId) {
        errs.push(_err('', '/deleteSurface/surfaceId', 'surfaceId is required'))
      }
      break
    }
  }

  return errs
}

// ── Surface Tree 校验 ─────────────────────────────

/**
 * 校验 surface 组件树的完整性。
 * 检查 root 存在、组件类型、dangling ref、cycle。
 */
export function validateSurfaceTree(srf: SurfaceState, cat?: Catalog): ValidationError[] {
  const errs: ValidationError[] = []

  // root 必须存在
  if (!srf.components.has(RootComponentID)) {
    errs.push(_err(srf.id, '', `surface has no "${RootComponentID}" component`))
  }

  // 检查每个组件
  for (const [id, comp] of srf.components) {
    if (cat && comp.type && !cat.components[comp.type]) {
      errs.push(_err(srf.id, '/' + id, `unknown component type "${comp.type}"`))
    }
    for (const ref of childRefsOf(comp, srf.catalogId)) {
      if (!srf.components.has(ref)) {
        errs.push(_err(srf.id, '/' + id, `references undefined component "${ref}"`))
      }
    }
  }

  // 循环检测（DFS）
  errs.push(..._detectCycles(srf))

  return errs
}

// ── 循环检测 ──────────────────────────────────────

enum Color { White, Gray, Black }

function _detectCycles(srf: SurfaceState): ValidationError[] {
  const color = new Map<string, Color>()
  const errs: ValidationError[] = []

  for (const id of srf.components.keys()) {
    color.set(id, Color.White)
  }

  const visit = (id: string) => {
    const comp = srf.components.get(id)
    if (!comp) return
    color.set(id, Color.Gray)
    for (const ref of childRefsOf(comp, srf.catalogId)) {
      switch (color.get(ref)) {
        case Color.Gray:
          errs.push(_err(srf.id, '/' + id, `circular reference involving "${ref}"`))
          break
        case Color.White:
          visit(ref)
          break
      }
    }
    color.set(id, Color.Black)
  }

  if (srf.components.has(RootComponentID)) {
    visit(RootComponentID)
  }

  // 也检查不在 root 子树中的组件
  for (const id of srf.components.keys()) {
    if (color.get(id) === Color.White) {
      visit(id)
    }
  }

  return errs
}
